File: src/lib/catalog/legendaryArmor.ts
import {
  asRawEquipmentSlot,
  asRawInventoryItem,
} from "@/lib/manifest/extractors/rawTypes";
import {
  getRaw,
  isUsable,
  projectBase,
  toArmorSlot,
  toClassName,
} from "@/lib/manifest/extractors/common";
import type { SetBonusRecord } from "@/lib/manifest/types/records";
import type { ManifestService } from "@/lib/manifest/types/services";

export type LegendaryArmorRow = {
  hash: number;
  name: string;
  searchName: string;
  icon: string | null;
  slot: string;
  classType?: string;
  setBonusName: string;
  setBonusHash: number;
  setBonusIcon?: string | null;
  setBonusPerks?: Array<{
    requiredCount: number;
    name: string;
    description: string;
  }>;
};

export type LegendaryArmorItemProjection = {
  name: string;
  searchName: string;
  icon: string | null;
  slot: string;
  classType?: string;
};

async function buildSlotMap(
  manifest: ManifestService,
  version: string,
): Promise<Map<number, string>> {
  const slotTable = await manifest.loadRawTable(version, "DestinyEquipmentSlotDefinition");
  const map = new Map<number, string>();
  for (const v of Object.values(slotTable)) {
    const slot = asRawEquipmentSlot(v);
    if (!slot) continue;
    const name = toArmorSlot(slot.displayProperties.name);
    if (name) map.set(slot.hash, name);
  }
  return map;
}

export function buildLegendaryArmorRows(
  setBonuses: SetBonusRecord[],
  projectItem: (hash: number) => LegendaryArmorItemProjection | null,
): LegendaryArmorRow[] {
  const rows: LegendaryArmorRow[] = [];
  const seen = new Set<number>();

  for (const set of setBonuses) {
    for (const hash of set.itemHashes) {
      if (seen.has(hash)) continue;
      const projection = projectItem(hash);
      if (!projection) continue;
      seen.add(hash);
      rows.push({
        hash,
        name: projection.name,
        searchName: projection.searchName,
        icon: projection.icon,
        slot: projection.slot,
        classType: projection.classType,
        setBonusName: set.name,
        setBonusHash: set.hash,
        setBonusIcon: set.icon ?? null,
        setBonusPerks: set.perks.map((p) => ({
          requiredCount: p.requiredCount,
          name: p.name,
          description: p.description,
        })),
      });
    }
  }

  return rows;
}

export async function loadLegendaryArmorRows(
  setBonuses: SetBonusRecord[],
  manifest: ManifestService,
  version: string,
): Promise<LegendaryArmorRow[]> {
  const [itemTable, slotMap] = await Promise.all([
    manifest.loadRawTable(version, "DestinyInventoryItemDefinition"),
    buildSlotMap(manifest, version),
  ]);

  return buildLegendaryArmorRows(setBonuses, (hash) => {
    const raw = getRaw(itemTable, hash);
    const item = asRawInventoryItem(raw);
    if (!item || !isUsable(item)) return null;
    const base = projectBase(item);
    const slotHash = item.equippingBlock?.equipmentSlotTypeHash;
    const slot = slotHash != null ? slotMap.get(slotHash) : undefined;
    if (!slot) return null;
    const classType = toClassName(item.classType) ?? undefined;
    return {
      name: base.name,
      searchName: base.searchName,
      icon: base.icon,
      slot,
      classType,
    };
  });
}

File: flutter/packages/manifest/lib/src/extractors/legendary_armor.dart
import '../raw/raw_types.dart';
import '../types/records.dart';
import '../types/services.dart';
import '../types/stores.dart';
import 'common.dart';

const _itemTypeArmor = 2;
const _tierLegendary = 5;

/// Extract legendary armor definitions for catalog browse (DART-062).
///
/// Tier-5 armor with resolvable slot + class. Archetype plug is optional.
/// Set-bonus enrichment is intentionally out of scope (see spec A3).
class LegendaryArmorExtractor implements EntityExtractor {
  @override
  MvpStoreName get store => MvpStoreName.legendaryArmor;

  @override
  Future<List<Object>> extract(LoadRawTable loadTable) async {
    final itemTable = await loadTable('DestinyInventoryItemDefinition');
    final slotTable = await loadTable('DestinyEquipmentSlotDefinition');
    final slotMap = _buildSlotMap(slotTable);
    final result = <LegendaryArmorRecord>[];

    for (final item in iterItems(itemTable)) {
      if (!_isLegendaryArmor(item)) continue;
      final slot = _resolveArmorSlot(item, slotMap);
      if (slot == null) continue;
      final classType = toClassName(item.classType);
      if (classType == null) continue;

      final archetypePlug = findSocketPlug(item, itemTable, (plug) {
        final cat = plug.plug?['plugCategoryIdentifier'] as String? ?? '';
        final typeName = plug.itemTypeDisplayName ?? '';
        return cat.contains('armor_archetypes') ||
            typeName.contains('Archetype');
      });

      final base = projectBase(item);
      result.add(
        LegendaryArmorRecord(
          hash: base.hash,
          name: base.name,
          searchName: base.searchName,
          icon: base.icon,
          classType: classType,
          slot: slot,
          archetype: archetypePlug?.displayProperties.name,
        ),
      );
    }
    return result;
  }
}

bool _isLegendaryArmor(RawInventoryItem item) {
  final tier = item.inventory?['tierType'] as num?;
  return item.itemType == _itemTypeArmor && tier?.toInt() == _tierLegendary;
}

Map<int, ArmorSlotName> _buildSlotMap(RawTable slotTable) {
  final map = <int, ArmorSlotName>{};
  for (final v in slotTable.values) {
    final slot = RawEquipmentSlot.tryParse(v);
    if (slot == null) continue;
    final name = toArmorSlot(slot.displayProperties.name);
    if (name != null) map[slot.hash] = name;
  }
  return map;
}

ArmorSlotName? _resolveArmorSlot(
  RawInventoryItem item,
  Map<int, ArmorSlotName> slotMap,
) {
  final h = item.equippingBlock?['equipmentSlotTypeHash'] as num?;
  if (h == null) return null;
  return slotMap[h.toInt()];
}
